### PR TITLE
 Fixed the windows line endings in the CSVCallback.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
### Summary

Thanks @huyu398 for your work!
This PR is an update of #10609. See #10609 for a description of the issue.

Here is #10609 with modifications according to the review. 
I have tested the code locally on a windows and ubuntu machine. It worked in both. 

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
Edit: I tested with python 3.

**Comment by gabrieldemarmiesse** [comment]
It seems there are non-trivial issues with strings encodings and python 2/3. This is sadly not my area of expertise. Any hint would be appreciated. For info, the error is:

```
self = <csv.DictWriter instance at 0x7f22b09cc878>
rowdict = {'epoch': 'epoch', 'loss': 'loss', 'val_loss': 'val_loss'}
    def writerow(self, rowdict):
>       return self.writer.writerow(self._dict_to_list(rowdict))
E       TypeError: write() argument 1 must be unicode, not str
```

**Comment by fchollet** [comment]
I believe that for Python2 you'll need to wrap the contents of the write in a `unicode()`  statement. There may be a better solution, though, that would work the same way for both versions.

**Comment by gabrieldemarmiesse** [comment]
It was tested also on windows with 3.6. I didn't check on windows with python 2.7.

File: keras/callbacks.py [hunk 1]
```diff
@@ -1130,9 +1132,17 @@ def on_train_begin(self, logs=None):
...
                 with open(self.filename, 'r' + self.file_flags) as f:
                     self.append_header = not bool(len(f.readline()))
-            self.csv_file = open(self.filename, 'a' + self.file_flags)
+            mode = 'a'
         else:
-            self.csv_file = open(self.filename, 'w' + self.file_flags)
+            mode = 'w'
+
+        if sys.version_info[0] < 3:
+            self.csv_file = io.open(self.filename,
```
Comment (Dref360) [inline keras/callbacks.py]: B is added in the constructor if we remove the NT check.

File: keras/callbacks.py [hunk 2]
```diff
@@ -1156,9 +1166,12 @@ def handle_value(k):
         if not self.writer:
             class CustomDialect(csv.excel):
                 delimiter = self.sep
-
+            fieldnames = ['epoch'] + self.keys
+            if sys.version_info[0] < 3:
```
Comment (Dref360) [inline keras/callbacks.py]: You can use six.PY2

**Comment by gabrieldemarmiesse** [comment]
Thanks for the help @Dref360

**Comment by fchollet** [comment]
Tests are failing.

**Pull request merged**